### PR TITLE
Url shortener

### PR DESCRIPTION
**Pull request by nehaparmar99**
## Issue that this pull request solves
 Link shortener service
Made under DWOC and SWOC
 Closes: #153 

### Brief description of what is fixed or changed
Added a link shortener service which accepts a query parameter link and shortens it to a code of length 5. One using that link is redirected to the original link if valid, else an error is thrown.

## Types of changes
I've added the API testing file in the url folder itself named API Tests.postman_collection.json
Also the API Tests are available at https://www.getpostman.com/collections/69d0f7207dd18940e91a

_Put an `x` in the boxes that apply_

- [x] New feature (non-breaking change which adds functionality)
## Checklist

_Put an `x` in the boxes that apply_

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] My changes does not break the current system and it passes all the current test cases.

### CODE REVIEW

**Comment by jackfrost13** [comment]
will try to have a look tomorrow
detailed review will be done mostly by Saturday

**Comment by jackfrost13** [comment]
Working on it

File: backend/index.js [hunk 1]
```diff
@@ -25,6 +25,9 @@ app.use(bodyParser.urlencoded({ limit: '50mb', extended: true }));
 // Response time
 app.use(responseTime({ suffix: false }));
 
+//middleware
+app.use(express.json({ extended: false }));
```
Comment (jackfrost13) [inline backend/index.js:29]: Why do you need this?

File: backend/app/routes/url/redirecturl.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+const Url = require('../../models/Url');
+
+module.exports = async (req, res) => {
+  try {
```
Comment (jackfrost13) [inline backend/app/routes/url/redirecturl.js:4]: We prefer using await-to-js. Please refactor

File: backend/app/routes/url/redirecturl.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    const url = await Url.findOne({ urlCode: req.params.code });
+    console.log(url);
+    if (url) {
+      return res.status(200).json({ url: url.longUrl });
+    } else {
+      return res.status(404).json('No url found');
```
Comment (jackfrost13) [inline backend/app/routes/url/redirecturl.js:10]: Please follow Error handling as we have implemented it in other routes

File: backend/app/routes/url/redirecturl.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const Url = require('../../models/Url');
+
+module.exports = async (req, res) => {
+  try {
+    const url = await Url.findOne({ urlCode: req.params.code });
+    console.log(url);
+    if (url) {
+      return res.status(200).json({ url: url.longUrl });
```
Comment (jackfrost13) [inline backend/app/routes/url/redirecturl.js:8]: You should redirect to the long url

File: backend/app/routes/url/redirecturl.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const Url = require('../../models/Url');
+
+module.exports = async (req, res) => {
+  try {
+    const url = await Url.findOne({ urlCode: req.params.code });
+    console.log(url);
```
Comment (jackfrost13) [inline backend/app/routes/url/redirecturl.js:6]: commits should be free of console logs

File: backend/app/routes/url/redirecturl.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    if (url) {
+      return res.status(200).json({ url: url.longUrl });
+    } else {
+      return res.status(404).json('No url found');
+    }
+  } catch (error) {
+    console.log(err);
+    res.status(500).json('server error');
```
Comment (jackfrost13) [inline backend/app/routes/url/redirecturl.js:14]: please follow our coding practices for error handling

File: backend/app/routes/url/shorturl.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        await url.save();
+        res.json(url);
+      }
+    } catch (err) {
+      console.log(err);
+      res.status(500).json('Server error');
+    }
+  } else {
+    //if longurl is not valid
+    res.status(401).json('Invalid long url');
```
Comment (jackfrost13) [inline backend/app/routes/url/shorturl.js:39]: please check our error handling and refactor

File: backend/app/routes/url/shorturl.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+          longUrl,
+          shortUrl,
+          urlCode: urlCode.substring(0, 5),
+          date: new Date(),
+        });
+        await url.save();
+        res.json(url);
+      }
+    } catch (err) {
+      console.log(err);
```
Comment (jackfrost13) [inline backend/app/routes/url/shorturl.js:34]: refactor error handling

File: backend/app/routes/url/shorturl.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+const validurl = require('valid-url');
+const shortid = require('shortid');
+const config = require('../../../config');
+const Url = require('../../models/Url');
+module.exports = async (req, res) => {
+  const { longUrl } = req.body;
+  const baseUrl = config.baseUrl;
+  // Check base url
+  if (!validurl.isUri(baseUrl)) {
```
Comment (jackfrost13) [inline backend/app/routes/url/shorturl.js:9]: How is this related to the user?

File: backend/app/routes/url/shorturl.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    return res.status(401).json('Invalid base url');
+  }
+  //Creating url code
+  const urlCode = shortid.generate();
+  //Check long url
+  if (validurl.isUri(longUrl)) {
```
Comment (jackfrost13) [inline backend/app/routes/url/shorturl.js:15]: you should check this using validation schema
Please refer to other routes for ref

File: backend/app/routes/url/shorturl.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    console.log('present');
+    try {
```
Comment (jackfrost13) [inline backend/app/routes/url/shorturl.js:17]: use await-to-js

File: backend/app/routes/url/shorturl.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+const Url = require('../../models/Url');
+module.exports = async (req, res) => {
+  const { longUrl } = req.body;
+  const baseUrl = config.baseUrl;
+  // Check base url
+  if (!validurl.isUri(baseUrl)) {
+    return res.status(401).json('Invalid base url');
+  }
+  //Creating url code
+  const urlCode = shortid.generate();
```
Comment (jackfrost13) [inline backend/app/routes/url/shorturl.js:13]: try using `nano id` instead of shortid, its faster and more reliable and you need not use substring later

**Review by jackfrost13 — CHANGES_REQUESTED** [review]
Please check my comments
Apologies for delayed response

**Comment by jackfrost13** [comment]
@nehaparmar99 Are you still working on this? Please update your status

**Comment by nehaparmar99** [comment]
> @nehaparmar99 Are you still working on this? Please update your status

Yes I am, I'll do it by tomorrow latest

**Comment by nehaparmar99** [comment]
@jackfrost13 I've made the changes.Please check it out.
Changes made:
1.Refactored error handling.
2.Nanoid used.
3.The shortUrl obtained now redirects to the Original long url

**Comment by nehaparmar99** [comment]
@jackfrost13 please review it

**Comment by jackfrost13** [comment]
Extremely sorry, slipped out of my review list.
Please make a habbit to request for re-review else we don't get notification

**Comment by jackfrost13** [comment]
@nehaparmar99 Hi, can you join for 15 mins meet? I can request for changes but since you have been blocked on this for a couple of days I have made changes and was thinking to help you on it

**Review by jackfrost13 — APPROVED** [review]

**Comment by jackfrost13** [comment]
Closes #153 
@nehaparmar99 Let me know if want a sync up on brief understanding as to why my changes are helpful, I will be available only on weekends

**Comment by jackfrost13** [comment]
Hi @nehaparmar99, we didn't hear back from you, please let me know if you had gone through my changes. The whole point of running this cohort sprint is to spread awareness and exchange knowledge and if you are not benefitting out of it then it defeats the whole purpose. We try our best to rely on high standard.

**Comment by nehaparmar99** [comment]
> Hi @nehaparmar99, we didn't hear back from you, please let me know if you had gone through my changes. The whole point of running this cohort sprint is to spread awareness and exchange knowledge and if you are not benefitting out of it then it defeats the whole purpose. We try our best to rely on high standard.

Hey I'm so sorry for not getting back.I had gone through the changes, and understood the point of them.I have been extremely busy with my college schedule, so couldn't find much time.I am so sorry for the delayed response.

**Pull request merged**